kms-message/src/kms_kmip_request.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kms_kmip_reader_writer_private.h"
+#include "kms_message/kms_kmip_request.h"
+#include "kms_message_private.h"
+
+#include <inttypes.h>
+#include <stdint.h>
+
+static void
+copy_writer_buffer (kms_request_t *req, kmip_writer_t *writer) {
+   const uint8_t *buf;
+   size_t buflen;
+
+   buf = kmip_writer_get_buffer (writer, &buflen);
+   req->kmip.data = malloc (buflen);
+   memcpy (req->kmip.data, buf, buflen);
+   req->kmip.len = (uint32_t) buflen;
```

Comment:
Note: I would like narrowing casts to be preceeded by a check or assertion to avoid unexpected implementation-defined runtime behavior, but this may be better suited for a project-wide audit of narrowing, signed, and implicit conversions, likely as part of a warning reduction effort.